File: src/leetcode/LongestConsecutive.java
package leetcode;

import java.util.*;

public class LongestConsecutive {
    /**
     * Time Complexity: O(n^2)
     * Space Complexity: O(n)
     */
    public static int longestConsecutiveWithBruteforce(int[] nums) {
        int answer = 0;
        Set<Integer> set = new HashSet<>();

        for (int num: nums) {
            set.add(num);
        }

        for (int num: nums) {
            if (!set.contains(num - 1)) {
                int length = 0;
                while (set.contains(num + length)) {
                    length++;
                }
                answer = Math.max(answer, length);
            }
        }

        return answer;
    }

    public static int longestConsecutiveWithSort(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }

        Arrays.sort(nums);

        Set<Integer> set = new HashSet<>();
        List<Integer> uniqueNums = new ArrayList<>();
        for (int num: nums) {
            if (!set.contains(num)) {
                set.add(num);
                uniqueNums.add(num);
            }
        }

        int answer = 0;
        int length = 0;
        for (int i = 1; i < uniqueNums.size(); i++) {
            if (uniqueNums.get(i) - uniqueNums.get(i - 1) == 1) {
                length++;
            } else {
                answer = Math.max(answer, length + 1);
                length = 0;
            }
        }

        return Math.max(answer, length + 1);
    }

    public static int longestConsecutiveWithHashSetOptimization(int[] nums) {
        Set<Integer> set = new HashSet<>();
        for (int num: nums) {
            set.add(num);
        }

        int answer = 0;
        for (int num: nums) {
            if (!set.contains(num - 1)) {
                int curNum = num;
                int length = 1;
                while (set.contains(curNum + 1)) {
                    curNum++;
                    length++;
                }
                answer = Math.max(answer, length);
            }
        }

        return answer;
    }

    public static void main(String[] args) {
        int[] nums1 = new int[]{100,4,200,1,3,2};
        System.out.println(longestConsecutiveWithBruteforce(nums1));
        System.out.println(longestConsecutiveWithSort(nums1));
        System.out.println(longestConsecutiveWithHashSetOptimization(nums1));

        int[] nums2 = new int[]{0,3,7,2,5,8,4,6,0,1};
        System.out.println(longestConsecutiveWithBruteforce(nums2));
        System.out.println(longestConsecutiveWithSort(nums2));
        System.out.println(longestConsecutiveWithHashSetOptimization(nums2));
    }
}

File: src/leetcode/kotlin/LongestConsecutive.kt
package leetcode.kotlin

class LongestConsecutive {
    private fun arrayContains(nums: IntArray, target: Int): Boolean {
        for (num in nums) {
            if (num == target) {
                return true
            }
        }
        return false
    }

    /**
     * Bruteforce
     * Time complexity: O(n^3)
     * Space complexity: O(1)
     */
    fun solution1(nums: IntArray): Int {
        var answer = 0

        for (num in nums) {
            var currentNum = num
            var currentSteak = 1
            while (arrayContains(nums, currentNum)) {
                currentNum++
                currentSteak++
            }

            answer = Math.max(answer, currentSteak)
        }

        return answer
    }

    /**
     * Sort
     * Time complexity: O(n * logn)
     * Space Complexity: O(1) or O(n)
     */
    fun solution2(nums: IntArray): Int {
        if (nums.isEmpty()) {
            return 0
        }

        val sortedNums = nums.sorted()
        var answer = 1
        var currentSteak = 1
        for (i in 1..<sortedNums.size) {
            if (sortedNums[i] != sortedNums[i - 1]) {
                if (sortedNums[i] == sortedNums[i - 1] + 1) {
                    currentSteak++
                } else {
                    answer = Math.max(answer, currentSteak)
                    currentSteak = 1
                }
            }
        }

        return answer
    }

    /**
     * Optimized solution
     * Time complexity: O(n)
     * Space complexity: O(n)
     */
    fun solution3(nums: IntArray): Int {
        val setNums = mutableSetOf<Int>()

        for (num in nums) {
            setNums.add(num)
        }

        var answer = 1
        for (num in setNums) {
            if (!setNums.contains(num - 1)) {
                var currentNum = num
                var currentSteak = 1
                while (setNums.contains(currentNum + 1)) {
                    currentNum++
                    currentSteak++
                }

                answer = Math.max(answer, currentSteak)
            }
        }

        return answer
    }
}